mountpoint-s3/src/prefetch/caching_stream.rs

```diff
@@ -0,0 +1,422 @@
+        let block_size = self.cache.block_size();
+        assert!(block_size > 0);
+
+        let block_aligned_byte_range =
+            (block_range.start * block_size)..(block_range.end * block_size).min(range.object_size() as u64);
+
+        trace!(
+            ?key,
+            range =? block_aligned_byte_range,
+            original_range =? range,
+            "fetching data from client"
+        );
+        let get_object_result = match self
+            .client
+            .get_object(
+                &self.bucket,
+                key,
+                Some(block_aligned_byte_range),
+                Some(self.cache_key.etag.clone()),
+            )
+            .await
+        {
+            Ok(get_object_result) => get_object_result,
+            Err(e) => {
+                error!(key, error=?e, "GetObject request failed");
+                self.part_queue_producer
+                    .push(Err(PrefetchReadError::GetRequestFailed(e)));
+                return;
+            }
+        };
+
+        pin_mut!(get_object_result);
+        let mut block_index = block_range.start;
+        let mut buffer = ChecksummedBytes::default();
+        loop {
+            match get_object_result.next().await {
+                Some(Ok((offset, body))) => {
+                    trace!(offset, length = body.len(), "received GetObject part");
+
+                    // Split the body into blocks.
+                    let mut body: Bytes = body.into();
+                    while !body.is_empty() {
+                        let remaining = (block_size as usize).saturating_sub(buffer.len()).min(body.len());
+                        let chunk = body.split_to(remaining);
+                        if let Err(e) = buffer.extend(chunk.into()) {
+                            error!(key, error=?e, "Integrity check failed");
+                            self.part_queue_producer.push(Err(e.into()));
+                            return;
+                        }
+                        if buffer.len() < block_size as usize {
+                            break;
+                        }
+                        self.update_cache(block_index, &buffer);
+                        self.part_queue_producer
+                            .push(Ok(self.make_part(buffer, block_index, &range)));
+                        block_index += 1;
+                        buffer = ChecksummedBytes::default();
+                    }
+                }
+                Some(Err(e)) => {
+                    error!(key, error=?e, "GetObject body part failed");
+                    self.part_queue_producer
+                        .push(Err(PrefetchReadError::GetRequestFailed(e)));
+                    break;
+                }
+                None => {
+                    if !buffer.is_empty() {
+                        if buffer.len() + (block_index * block_size) as usize == range.object_size() {
+                            // Write last block to the cache.
```

Comment:
Took me a second to parse this: this only applies if we're at the end of the entire object, not the end of the request? What happens if the last part of this request isn't block-sized -- we just drop it on the floor? Isn't that going to break caching the first 8MB of an object, where the requests won't be nicely MB-aligned?

Reply:
It needs a comment. The `if` handles exactly that case: the last part of the request is not block-sized but it reaches the end of the object. We do want to cache the (smaller) block.
In practice, this is the only case. Requests either end on a block boundary or at the end of the object, because they come from `get_aligned_request`. But we do no enforce it.

Reply:
I also struggled to grok this bit at first, but wasn't sure how we can make it clearer.

In what cases do we need to worry about a non-empty buffer where its not the end of the object? Should we add `unreachable!` or a warning in this case?

Reply:
I think "requests either end on a block boundary or at the end of the object" is the part I was missing. Can we just assert that? Or at least `debug_assert!` it if we're paranoid about not crashing if we got the math wrong, since it is recoverable.

Reply:
Here we are handling the response from `GetObject` though, so I don't think an assertion is appropriate.

Reply:
I've come around: added comments and asserts to clarify what is in the buffer throughout the loop but also the assert on the last block. It can only fail if we introduce a logic bug in this function or if the client starts returning the wrong range.